### DISCUSSION TITLE
Running nicegui in parallel with other functions

### DISCUSSION THREAD

**Discussion by itman496** [zauberzeug/nicegui#294] [Q&A] [answered]
I'm sure its just because I'm a newbie, but:

Whenever I let nicegui run by itself as the only function, it works great.

But what I am trying to make is a web interface (think like a wifi router) for existing software.  But if I try to run anything after I call ui.run() it just completely gives up the ghost.

This includes calling threads.. Which is weird.

I want nicegui to run at the same time, by itself, as other threads and whatnot in my program.  So it can be a webui to change global variables, call functions, or whatever, or be untouched, while the main program does whatever it does.  Like how a wifi router does its thing, then you can log into the web interface, mess with it, or not, and it keeps working and doing its own thing the whole time.

What is the proper method to run stuff nicegui in the background while the rest of the program continues to execute?

Thanks for making this btw :)  I do not know how javascript or html works..

This works:
```py
from nicegui import app, ui
import threading
import time

testVar = 0

def changevar(data):
    global testVar
    if data == 'INCREMENT':
        testVar += 1
    if data == 'DECREMENT':
        testVar -= 1
    if data == 'RESET':
        testVar = 0

def loopfunction():
    while True:
        valuemonitor()
        time.sleep(1)

def valuemonitor():
    global testVar
    print(f'TESTVAR: {testVar}')
    return

def run_threaded(job_func):  # spawn a function on a new thread
    job_thread = threading.Thread(target=job_func)
    job_thread.start()
    return

ui.button('INCREASE', on_click=lambda: changevar('INCREMENT'))
ui.button('DECREASE', on_click=lambda: changevar('DECREMENT'))
ui.button('ZERO OUT', on_click=lambda: changevar('RESET'))

ui.run()
```

this does not, it launches a webpage but never finishes loading nicegui, but the thread runs and starts printing to console:

```py
from nicegui import app, ui
import threading
import time

testVar = 0

def changevar(data):
    global testVar
    if data == 'INCREMENT':
        testVar += 1
    if data == 'DECREMENT':
        testVar -= 1
    if data == 'RESET':
        testVar = 0

def loopfunction():
    while True:
        valuemonitor()
        time.sleep(1)

def valuemonitor():
    global testVar
    print(f'TESTVAR: {testVar}')
    return

def run_threaded(job_func):  # spawn a function on a new thread
    job_thread = threading.Thread(target=job_func)
    job_thread.start()
    return

ui.button('INCREASE', on_click=lambda: changevar('INCREMENT'))
ui.button('DECREASE', on_click=lambda: changevar('DECREMENT'))
ui.button('ZERO OUT', on_click=lambda: changevar('RESET'))

ui.run()

run_threaded(loopfunction())
```

**Comment by ghost** [comment]
Hi @itman496,

I'm not sure about the proper method to run stuff in the background, that probably depends on your use case / scale. But threads should be fine. And to help you out with the threading issue:

This happens because the function "`loopfunction()`" already executes before being passed as a variable. If you change `run_threaded(loopfunction())` to `run_threaded(loopfunction)` it should work. Notice the missing () in the second one. By removing the brackets you pass the function instead of executing the function. 

Any function you pass to `threading.Thread(target=job_func)` will be executed by `threading.Thread` in the thread it spawns. So you don't have to execute it yourself.

I hope this brief explanation makes sense to you!

**Reply by itman496** [reply]
Thank you so much! I haven't done python in a while and I'm having to re-learn all the little minutia.  Worked great!

**Reply by ghost** [reply]
> Thank you so much! I haven't done python in a while and I'm having to re-learn all the little minutia. Worked great!

Hey, no problem! But might I suggest marking the reaction by @falkoschindler  as an answer for future readers? Since his reaction explains the preferred method of running synchronised tasks.

**Comment by falkoschindler** [comment, marked as answer]
We usually prefer asyncio over threads. Let me give an example based on your (simplified) code:

```py
import asyncio
from nicegui import app, ui

testVar = 0

def changevar(increment):
    global testVar
    testVar += increment

async def monitor():
    while True:
        print(f'TESTVAR: {testVar}')
        await asyncio.sleep(1)

ui.button('INCREASE', on_click=lambda: changevar(1))
ui.button('DECREASE', on_click=lambda: changevar(-1))

app.on_startup(monitor)

ui.run()
```

With `app.on_startup` we start a new _task_ (instead of a _thread_) right when the app starts up. In order not to block the main thread while sleeping, we use `asyncio.sleep` instead of `time.sleep`. This way the UI remains responsive and other async functions can run in the meantime. And not using multithreading saves you the hustle on ensuring threadsafety throughout your application.

**Reply by itman496** [reply]
thank you very much!
